### PR TITLE
new and updated chains

### PR DESCRIPTION
**Pull request by adamdelsol**
Added:
Eth2
Agoric
Regen Network

Updated:
Juno

### CODE REVIEW

File: core/chains/ethereum2.go [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+)
+
+type Ethereum2Response struct {
```
Comment (xenowits) [inline core/chains/ethereum2.go:11]:
```suggestion
type Eth2Response struct {
```

File: core/chains/ethereum2.go [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+	Detail string `json:"detail"`
+}
+
+// https://api.rated.network/docs#/
+// the call used is for up to 100 entities but there are currently only 30
+// should update if operator count grows beyond 100
+//
+// voting powers were not included as the api does not provide this metric
+// and the ui does not display it
+func Ethereum2() (int, error) {
```
Comment (xenowits) [inline core/chains/ethereum2.go:45]: suggest: changing `Ethereum` to just `Eth`

File: core/chains/ethereum2.go [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+	nakamotoCoefficient := 1
+	url := ""
+
+	url = fmt.Sprintf("https://api.rated.network/v0/eth/operators?window=all&idType=entity&size=100")
+	resp, err := http.Get(url)
+	if err != nil {
+		errBody, _ := ioutil.ReadAll(resp.Body)
+		var errResp Ethereum2ErrorResponse
+		json.Unmarshal(errBody, &errResp)
+		log.Println(errResp.Detail)
```
Comment (xenowits) [inline core/chains/ethereum2.go:57]: do we want to keep this log?

File: core/chains/juno.go [hunk 4]
```diff
@@ -61,34 +56,26 @@ func Juno() (int, error) {
...
-	})
+	sort.Slice(votingPowers, func(i, j int) bool { return votingPowers[i] > votingPowers[j] })
 
-	totalVotingPower := utils.CalculateTotalVotingPowerBigNums(votingPowers)
+	totalVotingPower := utils.CalculateTotalVotingPower(votingPowers)
 	fmt.Println("Total voting power:", totalVotingPower)
 
-	// now we're ready to calculate the nakomoto coefficient
-	nakamotoCoefficient := utils.CalcNakamotoCoefficientBigNums(totalVotingPower, votingPowers)
+	// // now we're ready to calculate the nakomoto coefficient
```
Comment (xenowits) [inline core/chains/juno.go:77]:
```suggestion
	// now we're ready to calculate the nakomoto coefficient
```

File: core/chains/regen.go [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+		votingPowers = append(votingPowers, int64(val))
+	}
+
+	// need to sort the powers in descending order since they are in random order
+	sort.Slice(votingPowers, func(i, j int) bool { return votingPowers[i] > votingPowers[j] })
+
+	totalVotingPower := utils.CalculateTotalVotingPower(votingPowers)
+	fmt.Println("Total voting power:", totalVotingPower)
+
+	// // now we're ready to calculate the nakomoto coefficient
```
Comment (xenowits) [inline core/chains/regen.go:77]:
```suggestion
	// now we're ready to calculate the nakomoto coefficient
```

**Review by xenowits — APPROVED** [review]
LGTM 🚀

**Comment by xenowits** [comment]
Thank you for the contribution!! Cheers 🎉

File: core/chains/ethereum2.go (hunk 1 above)
Comment (adamdelsol) [reply core/chains/ethereum2.go:11]: done

File: core/chains/ethereum2.go (hunk 2 above)
Comment (adamdelsol) [reply core/chains/ethereum2.go:45]: done

File: core/chains/ethereum2.go [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+	// increments NC when the controllingPenetration is surpassed
+	for _, ele := range response.Data {
+		val := ele.NetworkPenetration
+		cumulativePenetration += val
+		if cumulativePenetration < controllingPenetration {
+			nakamotoCoefficient += 1
+		}
+	}
+
+	fmt.Println("The Nakamoto coeffiecient for ethereum2 is", nakamotoCoefficient)
```
[the remark this answers is not in the document]
Comment (adamdelsol) [reply core/chains/ethereum2.go:83]: done

File: core/chains/juno.go (hunk 4 above)
Comment (adamdelsol) [reply core/chains/juno.go:77]: done

File: core/chains/regen.go (hunk 5 above)
Comment (adamdelsol) [reply core/chains/regen.go:77]: done

File: core/chains/ethereum2.go (hunk 3 above)
Comment (adamdelsol) [reply core/chains/ethereum2.go:57]: removed

**Pull request merged**